File: contents/quantum_systems/code/c/energy.c
#include <stdbool.h>
#include <complex.h>
#include <string.h>
#include <math.h>

#include <fftw3.h>

void fft(double complex *x, int n, bool inverse) {
    double complex y[n];
    memset(y, 0, sizeof(y));
    fftw_plan p;

    if (inverse) {
        p = fftw_plan_dft_1d(n, (fftw_complex*)x, (fftw_complex*)y,
                             FFTW_BACKWARD, FFTW_ESTIMATE);
    } else {
        p = fftw_plan_dft_1d(n, (fftw_complex*)x, (fftw_complex*)y,
                             FFTW_FORWARD, FFTW_ESTIMATE);
    }

    fftw_execute(p);
    fftw_destroy_plan(p);

    for (size_t i = 0; i < n; ++i) {
        x[i] = y[i] / sqrt((double)n);
    }
}

double calculate_energy(double complex *wfc, double complex *h_r,
                        double complex *h_k, double dx, size_t size) {
    double complex wfc_k[size];
    double complex wfc_c[size];
    memcpy(wfc_k, wfc, sizeof(wfc_k));
    fft(wfc_k, size, false);

    for (size_t i = 0; i < size; ++i) {
        wfc_c[i] = conj(wfc[i]);
    }

    double complex energy_k[size];
    double complex energy_r[size];

    for (size_t i = 0; i < size; ++i) {
        energy_k[i] = wfc_k[i] * h_k[i];
    }

    fft(energy_k, size, true);

    for (size_t i = 0; i < size; ++i) {
        energy_k[i] *= wfc_c[i];
        energy_r[i] = wfc_c[i] * h_r[i] * wfc[i];
    }

    double energy_final = 0;

    for (size_t i = 0; i < size; ++i) {
        energy_final += creal(energy_k[i] + energy_r[i]);
    }

    return energy_final * dx;
}

File: contents/quantum_systems/code/go/energy.go
package main

import (
	"fmt"
	"github.com/runningwild/go-fftw/fftw"
	"math/cmplx"
)

func fft(x []complex128, inv bool) {
	y := fftw.NewArray(len(x))
	for i := 0; i < len(x); i++ {
		y.Set(i, x[i])
	}

	if inv {
		plan := fftw.NewPlan(y, y, fftw.Backward, fftw.Estimate)
		plan.Execute()
		plan.Destroy()
	} else {
		plan := fftw.NewPlan(y, y, fftw.Forward, fftw.Estimate)
		plan.Execute()
		plan.Destroy()
	}

	for i := 0; i < len(x); i++ {
		x[i] = y.At(i) / cmplx.Sqrt(complex(float64(len(x)), 0))
	}

}

func calculateEnergy(wfc, h_r, h_k []complex128, dx float64) float64 {
	wfc_k := make([]complex128, len(wfc))
	wfc_c := make([]complex128, len(wfc))

	copy(wfc_k, wfc)
	fft(wfc_k, false)

	for i := 0; i < len(wfc); i++ {
		wfc_c[i] = cmplx.Conj(wfc[i])
	}

	energy_r := make([]complex128, len(wfc))
	energy_k := make([]complex128, len(wfc))

	for i := 0; i < len(wfc); i++ {
		energy_k[i] = wfc_k[i] * h_k[i]
	}

	fft(energy_k, true)

	for i := 0; i < len(wfc); i++ {
		energy_k[i] *= wfc_c[i]
		energy_r[i] = wfc_c[i] * h_r[i] * wfc[i]
	}

	energyFinal := 0.0

	for i := 0; i < len(wfc); i++ {
		energy_final += real(energy_k[i] + energy_r[i])
	}

	return energy_final * dx
}
